espnet2/tasks/abs_task.py

```diff
@@ -840,14 +885,75 @@ def main(cls, args: argparse.Namespace = None, cmd: Sequence[str] = None) -> Non
                 nbest=args.keep_n_best_checkpoints,
             )
 
+    @classmethod
+    def build_iter_factory(
+        cls,
+        args: argparse.Namespace,
+        data_path_and_name_and_type,
+        shape_files: Union[Tuple[str, ...], List[str]],
+        batch_type: str,
+        batch_size: int,
+        sort_in_batch: Optional[str],
+        train: bool = True,
+        num_batches: int = None,
+        distributed: bool = False,
+    ) -> AbsIterFactory:
+        assert check_argument_types()
+        if args.train_dtype in ("float32", "O0", "O1", "O2", "O3"):
+            dtype = "float32"
+        else:
+            dtype = args.train_dtype
+
+        dataset = ESPnetDataset(
+            data_path_and_name_and_type,
+            float_dtype=dtype,
+            preprocess=cls.build_preprocess_fn(args, train),
+        )
+        cls.check_task_requirements(dataset, args.allow_variable_data_keys)
+
+        if train:
+            num_iters_per_epoch = args.num_iters_per_epoch
+            shuffle = True
+        else:
+            num_iters_per_epoch = None
+            shuffle = False
+
+        batch_sampler = build_batch_sampler(
+            type=batch_type,
+            shape_files=shape_files,
+            max_lengths=args.max_length,
+            batch_size=batch_size,
+            sort_in_batch=sort_in_batch,
+            sort_batch=args.sort_batch,
+        )
+
+        batches = list(batch_sampler)
+        if num_batches is not None:
+            batches = batches[:num_batches]
+
+        if distributed:
+            world_size = torch.distributed.get_world_size()
+            rank = torch.distributed.get_rank()
+            batches = [batch[rank::world_size] for batch in batches]
```

Comment:
About distributed training: Note that we can't use pytorch DistributedSampler because we need custom batch sampler to implement variable batch-size.
Indeed, DistributedSampler simply split the samples by `world_size`,

https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py

so instead of it, split samples here.